design-proposals/vm-rollout-strategy/vm-rollout-strategy.md

```diff
@@ -0,0 +1,207 @@
+# Overview
+
+Today changes to a VM object (i.e. changing VM.spec.template) get applied, but only take effect upon the next restart - we call these pending changes “staged”. This is a proposal to introduce a custer-level rollout strategy field to control how changes to the VM object will propagate to the VirtualMachineInstance object without interupting the workload.
+
+
+## Motivation
+
+In most cases, the user would prefer the changes to the VM object to eventually propagate to the VMI object, as long as the rollout does not interrupt the workload. This behavior would be similar to how the Deployments roll out ReplicaSets.
+
+
+## Goals
+
+-   Provide a way to eventually propagate changes from a running VM object to its corresponding VMI without interrupting the workload.
+    
+-   Maintain backwards compatibility for cluster owners that prefer keeping staged changes until next VM reboot.
+    
+
+## Non Goals
+
+-   Live uppdate for **all** fields of the Virtual Machine API. Only select fields can be propagated to the guest. 
+-   Make LiveUpdate the default. We defer this decision.
+-   Support upgrades of VMs that are using the `liveUpdateFeatures`.
+-   This design does consider the hot-plugging of network interfaces and additional storage.
```

Comment:
The proposal is generic in all aspects until this non-goal.
Networking and storage hotplug exist, I think it will be beneficial to consider them as part of this proposal. The other alternative is to change this proposal naming and goals to be specific to two live changes (CPU & Memory).

I would prefer to keep it generic and pave the way for the harmonization of hotplug now.

Reply:
Ack, on keeping it generic - the API should be consistent regardless of the device that is hotplugged.

In addition this could be like in virsh, where you can define if it is for the definition, Live or both. So with that you could have a "non-persistent" (or ephemeral) hotplug as well.

Reply:
@EdDev do you have a constructive suggestion on how to make this proposal cover network, too? Would dropping this line be enough, or any other changes are suggested?

Reply:
> @EdDev do you have a constructive suggestion on how to make this proposal cover network, too? Would dropping this line be enough, or any other changes are suggested?

Yes, the suggestions have been added to the relevant threads.
https://github.com/kubevirt/community/pull/242/files#r1425171177

https://github.com/kubevirt/community/pull/242/files#r1425166735

Reply:
@EdDev If I read your replies correctly, you suggest that networking has a way forward and can be adjusted to this proposal.
The second comment is about a generic issue with the rollout dynamics, not specific to networking. 

Regarding storage, I wouldn't commit to it at this point. Storage hot plug is done today via a subresource. We need a separate design about how it can be implemented in a declarative manner.

Reply:
> @EdDev If I read your replies correctly, you suggest that networking has a way forward and can be adjusted to this proposal.

Yes, networking can confirm to the suggested strategy field when available.
We can update the network hotplug proposal/user-guide to reflect that before graduation.

> The second comment is about a generic issue with the rollout dynamics, not specific to networking.

It is indeed generic, but it effects networking. We want to make sure the proposal here is not limiting in the sense that network hotplug will depend on other changes done on the spec (e.g. CPU or Memory hotplug).
 
> Regarding storage, I wouldn't commit to it at this point. Storage hot plug is done today via a subresource. We need a separate design about how it can be implemented in a declarative manner.

I agree it needs a dedicated proposal for storage. We just need to leave this proposal in a state it can also be applied to storage in the future. This is why I think someone from the sig-storage should  also review this PR.

Reply:
Note: Storage has some problems concerning declarative hotplug because there is a dependency on the order of volumes.